### PR TITLE
Convert external int fields to uint32 or int32 as appropriate

### PR DESCRIPTION
**Pull request by smarterclayton**
Downgrades prober to int32 externally (not waiting 40 years for a pod to be ready)

Fuzzer is currently broken (internal int is not a valid fuzz target) which has to be fixed prior to merging this.

Addresses some possible problems from #17095.  Main motivation is that this is required for protobuf support because we can't have ambiguous external types.

### CODE REVIEW

**Comment by smarterclayton** [comment]
@lavalamp a custom fuzzer func for *int and **int doesn't seem to be invoked - do you know why a custom fuzzer func on a builtin type wouldn't be invoked?

**Comment by pmorie** [comment]
@pweil- who will make the changes in SecurityContextConstraints for this?

**Comment by pweil-** [comment]
> @pweil- who will make the changes in SecurityContextConstraints for this?

most likely me

**Comment by smarterclayton** [comment]
I'm making all these changes in my own PR

**Comment by smarterclayton** [comment]
Unless it hasn't been merged, in which case yes, that's someone else's
problem ;)

**Comment by smarterclayton** [comment]
@lavalamp disregard the fuzz problem, it's because goreflect returns "int64" for "int" in some contexts and "int" in others.  Still trying to figure out why.

**Comment by lavalamp** [comment]
That's weird.

File: pkg/util/util.go [hunk 1]
```diff
@@ -139,7 +139,7 @@ func Until(f func(), period time.Duration, stopCh <-chan struct{}) {
 // accept a name or number.
 type IntOrString struct {
 	Kind   IntstrKind
-	IntVal int
+	IntVal int32
```
Comment (lavalamp) [inline pkg/util/util.go]: It is really weird to see code like int(blah.IntVal). I can live with it if I have to, but I'd prefer not to...

* Leave as int, return error if it's > 2B when marshalling? (UGH)
* Change name to Int32Val? 
* Add a `Int() int` function that does the cast for you?

File: pkg/util/util.go [hunk 2]
```diff
@@ -161,6 +161,10 @@ func NewIntOrStringFromString(val string) IntOrString {
 	return IntOrString{Kind: IntstrString, StrVal: val}
 }
 
+func (intstr *IntOrString) IntValue() int {
+	return int(intstr.IntVal)
```
Comment (lavalamp) [inline pkg/util/util.go]: Oh, this function already exists. Can we call it instead of casting?

Comment (smarterclayton) [inline pkg/util/util.go]: Call it everywhere, you mean?  Happy to.

**Comment by smarterclayton** [comment]
It was a bad fuzzer - we were recreating the fuzzer and it was non obvious
anywhere.

**Comment by smarterclayton** [comment]
This is ready for review (still verifying i didn't break any tests).  Goal was 100% no `int` or `uint` in public code, if there are deeper refactorings or other changes I can do those in a follow up.

**Comment by smarterclayton** [comment]
I bumped `Prober` fields to uint32 because we shouldn't be waiting multiple years for a pod to become live.

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
How can I kick unit/integration?

**Comment by liggitt** [comment]
@k8s-bot unit test this

File: pkg/api/types.go [hunk 3]
```diff
@@ -719,11 +719,11 @@ type Probe struct {
 	// The action taken to determine the health of a container
 	Handler `json:",inline"`
 	// Length of time before health checking is activated.  In seconds.
-	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
+	InitialDelaySeconds int `json:"initialDelaySeconds,omitempty"`
```
Comment (thockin) [inline pkg/api/types.go]: shouldn't we be crystallizing all 'int' to fixed-width?

File: pkg/api/types.go [hunk 4]
```diff
@@ -1051,7 +1051,7 @@ type PodSecurityContext struct {
 	// 3. The permission bits are OR'd with rw-rw----
 	//
 	// If unset, the Kubelet will not modify the ownership and permissions of any volume.
-	FSGroup *int64 `json:"fsGroup,omitempty"`
+	FSGroup *uint32 `json:"fsGroup,omitempty"`
```
Comment (thockin) [inline pkg/api/types.go]: @pmorie if we had a type for UnixUID and UnixGID this would be easier... :)

File: pkg/api/types.go (hunk 3 above)
Comment (liggitt) [inline pkg/api/types.go]: +1

Comment (liggitt) [inline pkg/api/types.go]: Oh, hrmm, this was internal… iirc, the ripple was huge

Comment (liggitt) [inline pkg/api/types.go]: @smarterclayton how many bullets do you feel like biting?

Comment (smarterclayton) [inline pkg/api/types.go]: Internal changes are massive.  If you hate me, make me do it, but in a follow up.

Comment (smarterclayton) [inline pkg/api/types.go]: These are 32 ext and 32 or 64 internal, which is ok for now.

**Comment by smarterclayton** [comment]
I can't see what is failing on unit/int - I have green locally

File: api/swagger-spec/v1.json [hunk 5]
```diff
@@ -13666,7 +13666,6 @@
      },
      "runAsUser": {
       "type": "integer",
-      "format": "int64",
```
Comment (liggitt) [inline api/swagger-spec/v1.json]: missing format?

File: pkg/api/types.go (hunk 3 above)
Comment (thockin) [inline pkg/api/types.go]: We should document that in the conventions.  It's obviously easier to use `int` internally.  Will the codegen converters convert int32 -> int ?

Comment (smarterclayton) [inline pkg/api/types.go]: Yes, and generated conversions now properly tolerate the change.

Changing Int internally rippled out to most code paths - if we document
int64 external requires int64 internal (until we nuke internal), we can
attack it in pieces.

On Nov 17, 2015, at 11:29 PM, Tim Hockin <notifications@github.com> wrote:

In pkg/api/types.go
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r45158886>:

> @@ -719,11 +719,11 @@ type Probe struct {
>  	// The action taken to determine the health of a container
>  	Handler `json:",inline"`
>  	// Length of time before health checking is activated.  In seconds.
> -	InitialDelaySeconds int64 `json:"initialDelaySeconds,omitempty"`
> +	InitialDelaySeconds int `json:"initialDelaySeconds,omitempty"`

We should document that in the conventions. It's obviously easier to use int
internally. Will the codegen converters convert int32 -> int ?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17353/files#r45158886>.

Comment (thockin) [inline pkg/api/types.go]: If teh conversions handle it, fine.  for internal it is a lot less terrible.

File: api/swagger-spec/v1.json (hunk 5 above)
Comment (smarterclayton) [inline api/swagger-spec/v1.json]: Swagger considers int32 the canonical "integer".

Comment (smarterclayton) [inline api/swagger-spec/v1.json]: And there is no support for uint in swagger

**Comment by smarterclayton** [comment]
@k8s-bot unit test this

File: api/swagger-spec/v1.json (hunk 5 above)
Comment (liggitt) [inline api/swagger-spec/v1.json]: > Swagger considers int32 the canonical "integer".
> And there is no support for uint in swagger

nice. the integer's integer.

**Comment by smarterclayton** [comment]
@k8s-oncall, is/was the Jenkins unit/integration bot broken?  Haven't been able to have it rerun the unit/integration test?  The normal 'unit test this' isn't working.

**Comment by gmarek** [comment]
@k8s-bot unit test this

**Comment by smarterclayton** [comment]
Maybe I got reset off the list of people authorized to ask for that?

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
ok to test

File: pkg/api/errors/errors.go [hunk 6]
```diff
@@ -226,7 +226,7 @@ func NewServerTimeout(kind, operation string, retryAfterSeconds int) error {
 		Details: &unversioned.StatusDetails{
 			Kind:              kind,
 			Name:              operation,
-			RetryAfterSeconds: retryAfterSeconds,
+			RetryAfterSeconds: int32(retryAfterSeconds),
```
Comment (pmorie) [inline pkg/api/errors/errors.go]: I am guessing the narrowing/widening conversions here are because you don't want to change the API all over town, or do you really prefer the API for this to be an int.

**Comment by smarterclayton** [comment]
This is green - generous review appreciated from any and all... :)

**Comment by lavalamp** [comment]
Will look shortly

File: contrib/mesos/pkg/service/endpoints_controller.go [hunk 7]
```diff
@@ -429,8 +429,8 @@ func findPort(pod *api.Pod, svcPort *api.ServicePort) (int, int, error) {
 		p := portName.IntVal
```
Comment (lavalamp) [inline contrib/mesos/pkg/service/endpoints_controller.go]: Call the function instead of casting p below?

File: pkg/api/errors/errors.go (hunk 6 above)
Comment (lavalamp) [inline pkg/api/errors/errors.go]: I don't expect it makes a difference for this field, but casting like this is not safe in general as it potentially hides a cross-platform bug.

File: pkg/api/errors/errors.go [hunk 8]
```diff
@@ -412,7 +412,7 @@ func SuggestsClientDelay(err error) (int, bool) {
 		if t.Status().Details != nil {
 			switch t.Status().Reason {
 			case unversioned.StatusReasonServerTimeout, unversioned.StatusReasonTimeout:
-				return t.Status().Details.RetryAfterSeconds, true
+				return int(t.Status().Details.RetryAfterSeconds), true
```
Comment (lavalamp) [inline pkg/api/errors/errors.go]: This direction, casting from int32 to int, should be safe everywhere.

File: pkg/api/testing/fuzzer.go [hunk 9]
```diff
@@ -45,6 +45,17 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 		f.RandSource(src)
 	}
 	f.Funcs(
+		func(j *int, c fuzz.Continue) {
+			*j = int(c.Int31())
+		},
+		func(j **int, c fuzz.Continue) {
```
Comment (lavalamp) [inline pkg/api/testing/fuzzer.go]: I'm surprised you need this function, it should be the default behavior.

File: pkg/api/testing/fuzzer.go [hunk 10]
```diff
@@ -138,8 +149,8 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 		},
 		func(j *extensions.JobSpec, c fuzz.Continue) {
 			c.FuzzNoCustom(j) // fuzz self without calling this function again
-			completions := c.Rand.Int()
-			parallelism := c.Rand.Int()
+			completions := int(c.Rand.Int31())
```
Comment (lavalamp) [inline pkg/api/testing/fuzzer.go]: The versioned struct is int32 and the internal is int?

File: pkg/api/testing/fuzzer.go [hunk 11]
```diff
@@ -361,9 +371,9 @@ func FuzzerFor(t *testing.T, version string, src rand.Source) *fuzz.Fuzzer {
 		},
 		func(s *extensions.HorizontalPodAutoscalerSpec, c fuzz.Continue) {
 			c.FuzzNoCustom(s) // fuzz self without calling this function again
-			minReplicas := c.Rand.Int()
+			minReplicas := int(c.Rand.Int31())
 			s.MinReplicas = &minReplicas
-			s.CPUUtilization = &extensions.CPUTargetUtilization{TargetPercentage: int(c.RandUint64())}
+			s.CPUUtilization = &extensions.CPUTargetUtilization{TargetPercentage: int(c.Int31())}
```
Comment (lavalamp) [inline pkg/api/testing/fuzzer.go]: This will no longer test negative numbers.

File: pkg/api/errors/errors.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/api/errors/errors.go]: Agree, but I'd rather refactor the internal in a follow on (remove casts to
int32)

File: pkg/api/testing/fuzzer.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/api/testing/fuzzer.go]: int is 64 bit internally, so fuzzInt generates a 64 bit number which won't
cross the boundary..  That won't round trip to the external types.  If /
when we refactor the internal, we would not need to do this.

File: pkg/api/testing/fuzzer.go (hunk 10 above)
Comment (smarterclayton) [inline pkg/api/testing/fuzzer.go]: Correct

File: pkg/api/testing/fuzzer.go (hunk 11 above)
Comment (smarterclayton) [inline pkg/api/testing/fuzzer.go]: Going to fix with int(int32(c.RandUint64()))

File: pkg/apis/extensions/v1beta1/defaults_test.go [hunk 12]
```diff
@@ -294,9 +294,9 @@ func roundTrip(t *testing.T, obj runtime.Object) runtime.Object {
 	return obj3
 }
 
-func newInt(val int) *int {
-	p := new(int)
-	*p = val
+func newInt(val int) *int32 {
```
Comment (lavalamp) [inline pkg/apis/extensions/v1beta1/defaults_test.go]: Fix function name?

File: pkg/util/intstr/intstr.go [hunk 13]
```diff
@@ -44,7 +44,7 @@ const (
 
 // FromInt creates an IntOrString object with an int value.
```
Comment (lavalamp) [inline pkg/util/intstr/intstr.go]: Add: it's your responsibility to not call this with an int that won't fit in 32 bits.

File: pkg/util/intstr/intstr.go [hunk 14]
```diff
@@ -67,7 +67,17 @@ func (intstr *IntOrString) String() string {
 	if intstr.Type == String {
 		return intstr.StrVal
 	}
-	return strconv.Itoa(intstr.IntVal)
+	return strconv.Itoa(intstr.IntValue())
+}
+
+// IntValue returns the IntVal cast as int32 if type Int, or if
```
Comment (lavalamp) [inline pkg/util/intstr/intstr.go]: s/32//

**Comment by lavalamp** [comment]
minor comments

I don't like that the conversion functions fail silently on ints bigger than 32 bits, but I was waiting to see a field for which it would matter to comment about it and I didn't find one. But this int/int32 casting business is a bad code smell.

LGTM

File: pkg/api/testing/fuzzer.go (hunk 9 above)
Comment (lavalamp) [inline pkg/api/testing/fuzzer.go]: Ah, I see.

**Comment by lavalamp** [comment]
If this wasn't such an extensive change, I'd ask for a squash. :)

**Comment by smarterclayton** [comment]
Trivial rebase, reapplying LGTM

**Comment by wojtek-t** [comment]
@smarterclayton - can you please rebase this PR?

**Comment by smarterclayton** [comment]
Trivial rebase, reapplying LGTM

**Comment by wojtek-t** [comment]
hmm - not sure what exactly happened, but it needs rebase again

**Comment by smarterclayton** [comment]
possible e2e flake on `create a job from an image when restart is OnFailure`

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
Trivial rebase, reapplying LGTM

File: pkg/util/intstr/intstr.go [hunk 15]
```diff
@@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string {
...
 		return intstr.StrVal
 	}
-	return strconv.Itoa(intstr.IntVal)
+	return strconv.Itoa(intstr.IntValue())
+}
+
+// IntValue returns the IntVal if type Int, or if
+// it is a String, will attempt a conversion to int.
+func (intstr *IntOrString) IntValue() int {
+	if intstr.Type == String {
```
Comment (thockin) [inline pkg/util/intstr/intstr.go]: I think this is wrong and we should not do it.  It's scope creep for this PR anyway.  Was there a use case justifying this?  I think we should roll it back

Comment (smarterclayton) [inline pkg/util/intstr/intstr.go]: It was to prevent having to refactor 900 lines of internal code that assume
IntOrString is int, not int32.  The string conversion bit can be removed as
long as we doc it (all the places it was used was the IntValue)

Comment (thockin) [inline pkg/util/intstr/intstr.go]: I don't mean the function, I mean the new cast of string -> int.  I assert
that this is a discriminated union, and not a "best effort to figure out
what the user meant".  if they said type String and read it back as an int,
they get a 0 or an error or a panic.  Not an atoi() of the string.

Comment (smarterclayton) [inline pkg/util/intstr/intstr.go]: I will nuke the best effort in a pr.

On Dec 1, 2015, at 1:38 AM, Tim Hockin <notifications@github.com> wrote:

In pkg/util/intstr/intstr.go
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46244484>:

> @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string {
>  	if intstr.Type == String {
>  		return intstr.StrVal
>  	}
> -	return strconv.Itoa(intstr.IntVal)
> +	return strconv.Itoa(intstr.IntValue())
> +}
> +
> +// IntValue returns the IntVal if type Int, or if
> +// it is a String, will attempt a conversion to int.
> +func (intstr *IntOrString) IntValue() int {
> +	if intstr.Type == String {

I don't mean the function, I mean the new cast of string -> int. I assert
that this is a discriminated union, and not a "best effort to figure out
what the user meant". if they said type String and read it back as an int,
they get a 0 or an error or a panic. Not an atoi() of the string.
… <#>
On Mon, Nov 30, 2015 at 11:44 AM, Clayton Coleman <notifications@github.com>
wrote: In pkg/util/intstr/intstr.go <#17353 (comment)
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46192602>>
: > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string { > if
intstr.Type == String { > return intstr.StrVal > } > - return
strconv.Itoa(intstr.IntVal) > + return strconv.Itoa(intstr.IntValue()) > +}
> + > +// IntValue returns the IntVal if type Int, or if > +// it is a
String, will attempt a conversion to int. > +func (intstr *IntOrString)
IntValue() int { > + if intstr.Type == String { It was to prevent having to
refactor 900 lines of internal code that assume IntOrString is int, not
int32. The string conversion bit can be removed as long as we doc it (all
the places it was used was the IntValue) … <#15159ed0968e77c8_> On Mon, Nov
30, 2015 at 1:27 PM, Tim Hockin <notifications@github.com> wrote: In
pkg/util/intstr/intstr.go <#17353
<https://github.com/kubernetes/kubernetes/pull/17353> (comment) <#17353
(comment)
<https://github.com/kubernetes/kubernetes/pull/17353#discussion_r46181719>>>
: > @@ -67,7 +71,17 @@ func (intstr *IntOrString) String() string { > if
intstr.Type == String { > return intstr.StrVal > } > - return
strconv.Itoa(intstr.IntVal) > + return strconv.Itoa(intstr.IntValue()) > +}
> + > +// IntValue returns the IntVal if type Int, or if > +// it is a
String, will attempt a conversion to int. > +func (intstr *IntOrString)
IntValue() int { > + if intstr.Type == String { I think this is wrong and
we should not do it. It's scope creep for this PR anyway. Was there a use
case justifying this? I think we should roll it back — Reply to this email
directly or view it on GitHub <
https://github.com/kubernetes/kubernetes/pull/17353/files#r46181719>. —
Reply to this email directly or view it on GitHub <
https://github.com/kubernetes/kubernetes/pull/17353/files#r46192602>.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17353/files#r46244484>.

**Comment by nikhiljindal** [comment]
@smarterclayton can you please add this to https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md as well?

**Comment by smarterclayton** [comment]
Weird, I thought we added it.  Yes

**Comment by smarterclayton** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/18941
